### PR TITLE
[#1339] feat(jdbc-doris): Support Apache Doris Catalog 

### PR DESCRIPTION
**Pull request by zhoukangcn**
### What changes were proposed in this pull request?

Support Apache Doris as a type of data source.

### Why are the changes needed?

Fix: #1339 

### Does this PR introduce _any_ user-facing change?

N/A

### How was this patch tested?

UT & IT

### CODE REVIEW

**Comment by yuqi1129** [comment]
@zhoukangcn 
Is it ready for review?

**Comment by zhoukangcn** [comment]
@yuqi1129 yes, please review this PR. Thanks.

File: catalogs/catalog-jdbc-doris/src/main/java/com/datastrato/gravitino/catalog/doris/operation/DorisTableOperations.java [hunk 1]
```diff
@@ -0,0 +1,619 @@
...
+  }
+
+  @Override
+  protected List<Index> getIndexes(String databaseName, String tableName, DatabaseMetaData metaData)
+      throws SQLException {
+    // get Indexes from SQL
+    try {
+      Connection connection = metaData.getConnection();
+
+      String sql = "SHOW INDEX FROM " + tableName + " FROM " + databaseName;
```
Comment (yuqi1129) [inline catalogs/catalog-jdbc-doris/src/main/java/com/datastrato/gravitino/catalog/doris/operation/DorisTableOperations.java:269]: Perhaps you need to backquote the `tableName` and `databaseName`

File: catalogs/catalog-jdbc-doris/src/main/java/com/datastrato/gravitino/catalog/doris/operation/DorisTableOperations.java [hunk 2]
```diff
@@ -0,0 +1,619 @@
...
+      PreparedStatement preparedStatement = connection.prepareStatement(sql);
+      ResultSet resultSet = preparedStatement.executeQuery();
+      List<Index> indexes = new ArrayList<>();
+      while (resultSet.next()) {
+        String indexName = resultSet.getString("Key_name");
+        String columnName = resultSet.getString("Column_name");
+        indexes.add(
+            Indexes.of(Index.IndexType.PRIMARY_KEY, indexName, new String[][] {{columnName}}));
```
Comment (yuqi1129) [inline catalogs/catalog-jdbc-doris/src/main/java/com/datastrato/gravitino/catalog/doris/operation/DorisTableOperations.java:277]: Does Doris only have the primary key?

File: catalogs/catalog-jdbc-doris/src/main/java/com/datastrato/gravitino/catalog/doris/operation/DorisTableOperations.java [hunk 3]
```diff
@@ -0,0 +1,619 @@
...
+      } else if (change instanceof TableChange.AddColumn) {
+        TableChange.AddColumn addColumn = (TableChange.AddColumn) change;
+        lazyLoadTable = getOrCreateTable(databaseName, tableName, lazyLoadTable);
+        alterSql.add(addColumnFieldDefinition(addColumn));
+      } else if (change instanceof TableChange.RenameColumn) {
+        throw new IllegalArgumentException("Rename column is not supported yet");
+      } else if (change instanceof TableChange.UpdateColumnType) {
+        lazyLoadTable = getOrCreateTable(databaseName, tableName, lazyLoadTable);
+        TableChange.UpdateColumnType updateColumnType = (TableChange.UpdateColumnType) change;
+        alterSql.add(updateColumnTypeFieldDefinition(updateColumnType, lazyLoadTable));
```
Comment (yuqi1129) [inline catalogs/catalog-jdbc-doris/src/main/java/com/datastrato/gravitino/catalog/doris/operation/DorisTableOperations.java:354]: Please review the backquote issues regarding column names, table names, etc in the whole PR carefully.

File: catalogs/catalog-jdbc-doris/src/main/java/com/datastrato/gravitino/catalog/doris/operation/DorisTableOperations.java [hunk 4]
```diff
@@ -0,0 +1,619 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import org.apache.commons.collections4.CollectionUtils;
+import org.apache.commons.lang3.BooleanUtils;
+import org.apache.commons.lang3.StringUtils;
+
+/** Table operations for Doris. */
+public class DorisTableOperations extends JdbcTableOperations {
+  public static final String BACK_QUOTE = "`";
```
Comment (diqiu50) [inline catalogs/catalog-jdbc-doris/src/main/java/com/datastrato/gravitino/catalog/doris/operation/DorisTableOperations.java:46]: Similarly, does the database name require the `BACK_QUOTE`

File: catalogs/catalog-jdbc-doris/src/main/java/com/datastrato/gravitino/catalog/doris/operation/DorisDatabaseOperations.java [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+      throw this.exceptionMapper.toGravitinoException(sqlException);
+    }
+
+    return sqlBuilder.toString();
+  }
+
+  @Override
+  public JdbcSchema load(String databaseName) throws NoSuchSchemaException {
+    try (final Connection connection = this.dataSource.getConnection()) {
+      String query = "SELECT * FROM information_schema.SCHEMATA WHERE SCHEMA_NAME = ?";
```
[the remark this answers is not in the document]
Comment (zhoukangcn) [reply catalogs/catalog-jdbc-doris/src/main/java/com/datastrato/gravitino/catalog/doris/operation/DorisDatabaseOperations.java:76]: @diqiu50 Doris is case-insensitive

[the remark this answers is not in the document]
Comment (zhoukangcn) [reply catalogs/catalog-jdbc-doris/src/main/java/com/datastrato/gravitino/catalog/doris/operation/DorisDatabaseOperations.java:76]: @yuqi1129 yes, user has default privilege to use database, and can list all tables.

File: catalogs/catalog-jdbc-doris/src/main/java/com/datastrato/gravitino/catalog/doris/operation/DorisTableOperations.java [hunk 6]
```diff
@@ -0,0 +1,619 @@
...
+        "Doris only supports HASH or EVEN distribution strategy");
+
+    if (distribution.strategy() == Strategy.HASH) {
+      // Check if the distribution column exists
+      Arrays.stream(distribution.expressions())
+          .forEach(
+              expression -> {
+                Preconditions.checkArgument(
+                    Arrays.stream(columns)
+                        .anyMatch(column -> column.name().equalsIgnoreCase(expression.toString())),
```
[the remark this answers is not in the document]
Comment (zhoukangcn) [reply catalogs/catalog-jdbc-doris/src/main/java/com/datastrato/gravitino/catalog/doris/operation/DorisTableOperations.java:192]: yes, Doris column name is case-insensitive, and will convert to lowercase

**Comment by zhoukangcn** [comment]
@yuqi1129 @diqiu50 
I have updated the PR, please review if you are free.

File: catalogs/catalog-jdbc-doris/src/main/java/com/datastrato/gravitino/catalog/doris/converter/DorisTypeConverter.java [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+        return Types.FloatType.get();
+      case DOUBLE:
+        return Types.DoubleType.get();
+      case DECIMAL:
+        return Types.DecimalType.of(
+            Integer.parseInt(typeBean.getColumnSize()), Integer.parseInt(typeBean.getScale()));
+      case DATE:
+        return Types.DateType.get();
+      case DATETIME:
+        return Types.TimeType.get();
```
Comment (mchades) [inline catalogs/catalog-jdbc-doris/src/main/java/com/datastrato/gravitino/catalog/doris/converter/DorisTypeConverter.java:48]: `DATETIME` should convert to `timestamp`?

File: catalogs/catalog-jdbc-doris/src/main/java/com/datastrato/gravitino/catalog/doris/converter/DorisTypeConverter.java [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+      return DECIMAL
+          + "("
+          + ((Types.DecimalType) type).precision()
+          + ","
+          + ((Types.DecimalType) type).scale()
+          + ")";
+    } else if (type instanceof Types.DateType) {
+      return DATE;
+    } else if (type instanceof Types.TimeType) {
+      return DATETIME;
```
Comment (mchades) [inline catalogs/catalog-jdbc-doris/src/main/java/com/datastrato/gravitino/catalog/doris/converter/DorisTypeConverter.java:87]: Wrong convert too?

File: catalogs/catalog-jdbc-doris/src/main/java/com/datastrato/gravitino/catalog/doris/operation/DorisDatabaseOperations.java [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+/** Database operations for Doris. */
+public class DorisDatabaseOperations extends JdbcDatabaseOperations {
+  private static final String BACK_QUOTE = "`";
+
+  @Override
+  public String generateCreateDatabaseSql(
+      String databaseName, String comment, Map<String, String> properties) {
+    if (StringUtils.isNotEmpty(comment)) {
+      LOG.warn(
+          "Ignoring comment option on database create. doris does not support comment option on database create.");
```
Comment (mchades) [inline catalogs/catalog-jdbc-doris/src/main/java/com/datastrato/gravitino/catalog/doris/operation/DorisDatabaseOperations.java:31]: `doris` -> `Doris`

File: catalogs/catalog-jdbc-doris/src/test/java/com/datastrato/gravitino/catalog/doris/utils/TestDorisUtils.java [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+public class TestDorisUtils {
+  @Test
+  public void generatePropertiesSql() {
+    // Test When properties is null
```
Comment (smarthanwang) [inline catalogs/catalog-jdbc-doris/src/test/java/com/datastrato/gravitino/catalog/doris/utils/TestDorisUtils.java:16]: typo: When -> when

**Pull request closed without merging**